src/modules/mc_rate_control/MulticopterRateControl.cpp

```diff
@@ -0,0 +1,542 @@
+			_rates_i_scaled = _rate_i.emult(pid_attenuations(_thrust_sp, _param_mc_tpa_break_i.get(), _param_mc_tpa_rate_i.get()));
+			_rates_d_scaled = _rate_d.emult(pid_attenuations(_thrust_sp, _param_mc_tpa_break_d.get(), _param_mc_tpa_rate_d.get()));
+
+			_thrust_sp_prev = _thrust_sp;
+		}
+
+		// calculate loop update rate while disarmed or at least a few times (updating the filter is expensive)
+		if (!_v_control_mode.flag_armed || (now - _task_start) < 3300000) {
+			_dt_accumulator += dt;
+			++_loop_counter;
+
+			if (_dt_accumulator > 1.0f) {
+				const float loop_update_rate = (float)_loop_counter / _dt_accumulator;
+				_loop_update_rate_hz = _loop_update_rate_hz * 0.5f + loop_update_rate * 0.5f;
+				_dt_accumulator = 0;
+				_loop_counter = 0;
+				_lp_filters_d.set_cutoff_frequency(_loop_update_rate_hz, _param_mc_dterm_cutoff.get());
+			}
+		}
+
+		// run the rate controller
+		if (_v_control_mode.flag_control_rates_enabled) {
+			control_attitude_rates(dt, rates);
+			publish_rate_controller_status();
+			publish_actuator_controls(angular_velocity.timestamp_sample);
+
+		} else if (_v_control_mode.flag_control_termination_enabled) {
+			if (!_vehicle_status.is_vtol) {
+				_rates_sp.zero();
+				_rates_int.zero();
+				_thrust_sp = 0.0f;
+				_att_control.zero();
+				publish_actuator_controls(angular_velocity.timestamp_sample);
+			}
+		}
+	}
+
+	perf_end(_loop_perf);
+}
+
+int MulticopterRateControl::task_spawn(int argc, char *argv[])
+{
+	MulticopterRateControl *instance = new MulticopterRateControl();
+
+	if (instance) {
+		_object.store(instance);
+		_task_id = task_id_is_work_queue;
+
+		if (instance->init()) {
+			return PX4_OK;
+		}
+
+	} else {
+		PX4_ERR("alloc failed");
+	}
+
+	delete instance;
+	_object.store(nullptr);
+	_task_id = -1;
+
+	return PX4_ERROR;
+}
+
+int MulticopterRateControl::print_status()
+{
+	PX4_INFO("Running");
+
+	perf_print_counter(_loop_perf);
+	perf_print_counter(_loop_interval_perf);
+	perf_print_counter(_input_latency_perf);
+
+	return 0;
+}
+
+int MulticopterRateControl::custom_command(int argc, char *argv[])
+{
+	return print_usage("unknown command");
+}
+
+int MulticopterRateControl::print_usage(const char *reason)
+{
+	if (reason) {
+		PX4_WARN("%s\n", reason);
+	}
+
+	PRINT_MODULE_DESCRIPTION(
+		R"DESCR_STR(
+### Description
+This implements the multicopter rate controller. It takes rate setpoints (in acro mode
+via `manual_control_setpoint` topic) as inputs and outputs actuator control messages.
+
+The controller has a PID loop for angular rate error.
+
+Publication documenting the implemented Quaternion Attitude Control:
+Nonlinear Quadrocopter Attitude Control (2013)
+by Dario Brescianini, Markus Hehn and Raffaello D'Andrea
+Institute for Dynamic Systems and Control (IDSC), ETH Zurich
+
+https://www.research-collection.ethz.ch/bitstream/handle/20.500.11850/154099/eth-7387-01.pdf
```

Comment:
The rate controller is "just" a PID, nothing specific to that publication, so we can remove the citation.

Reply:
Removed.